### PR TITLE
SALTO-6074 PagerDuty - resolve userId to email

### PR DESCRIPTION
**Pull request by Uri-Bar**
SALTO-6074 PagerDuty - resolve userId to email
---

_Additional context for reviewer_

---
_Release Notes_: 
none

---
_User Notifications_: 
none

### CODE REVIEW

File: packages/pagerduty-adapter/src/adapter_creator.ts [hunk 1]
```diff
@@ -57,6 +69,10 @@ export const adapter = createAdapter<Credentials, Options, UserConfig>({
     connectionCreatorFromConfig: () => createConnection,
     credentialsFromConfig: defaultCredentialsFromConfig,
     additionalChangeValidators: createChangeValidator,
+    customizeFilterCreators: args => ({
+      ...filters.createCommonFilters<Options, UserConfig>(args),
+      convertUsersIds,
```
Comment (edenhassid) [inline packages/pagerduty-adapter/src/adapter_creator.ts:74]: why is this separate and not part of all the filters?

File: packages/pagerduty-adapter/src/filters/users.ts [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+  const targets = _.get(value, 'targets')
+  return Array.isArray(targets) && targets.every(target => _.isString(target.id) && _.isString(target.type))
+}
+
+const isScheduleLayerUser = (value: unknown): value is ScheduleLayerUser => {
+  const user = _.get(value, 'user')
+  return _.isPlainObject(user) && _.isString(user.id) && _.isString(user.type)
+}
+
+const replaceValues = (instance: InstanceElement, mapping: Record<string, string>): void => {
```
Comment (edenhassid) [inline packages/pagerduty-adapter/src/filters/users.ts:51]: can you add a comment with an example for each type?

File: packages/pagerduty-adapter/src/filters/users.ts [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+      break
+
+    case SCHEDULE_LAYERS_TYPE_NAME:
+      makeArray(instance.value.users)
+        .filter(isScheduleLayerUser)
+        .forEach(userObj => {
+          const userIdentifier = _.get(userObj, 'user.id')
+          if (mapping[userIdentifier]) {
+            _.set(userObj, 'user.id', mapping[userIdentifier])
+          }
```
Comment (edenhassid) [inline packages/pagerduty-adapter/src/filters/users.ts:75]: maybe add a log if `mapping[userIdentifier]` is undefined? (same for both cases)

File: packages/pagerduty-adapter/src/filters/users.ts [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+        })
+      break
+    default:
+      break
+  }
+}
+
+export const replaceValuesForChanges = async (
```
Comment (edenhassid) [inline packages/pagerduty-adapter/src/filters/users.ts:83]: why do you export this?

File: packages/pagerduty-adapter/src/filters/users.ts [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+  return {
+    name: 'usersFilter',
+    onFetch: async elements => {
+      if (!(config.fetch.convertUsersIds ?? DEFAULT_CONVERT_USERS_IDS_VALUE)) {
+        log.debug('Converting user ids was disabled (onFetch)')
+        return
+      }
+
+      // Using casting as the difference between definitions type and RequiredDefinitions is that the fetch is not mandatory in  the definitions, but I override it.
+      const users = await fetchUtils.getElements({
```
Comment (edenhassid) [inline packages/pagerduty-adapter/src/filters/users.ts:115]: 1. do we really want to use getElements? isn't it enough just to have a map of user id to email?

2. also we would want this code to run only if we have relevant instances right?

File: packages/pagerduty-adapter/src/filters/users.ts [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+      if (!users || _.isEmpty(users)) {
+        log.warn('Could not find any users (onFetch)')
+      }
+
+      userIdToLogin = Object.fromEntries(
+        users.elements.filter(isInstanceElement).map(user => [user.value.id, user.value.email]),
+      )
+      const loginToUserId = Object.fromEntries(
+        users.elements.filter(isInstanceElement).map(user => [user.value.email, user.value.id]),
+      )
```
Comment (edenhassid) [inline packages/pagerduty-adapter/src/filters/users.ts:155]: I think you can put all this logic in the utils as well

File: packages/pagerduty-adapter/src/filters/users.ts [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+      await replaceValuesForChanges(relevantChanges, loginToUserId)
+    },
+    onDeploy: async (changes: Change<InstanceElement>[]) => {
```
Comment (edenhassid) [inline packages/pagerduty-adapter/src/filters/users.ts:158]: I am not sure we need the onDeploy actually, cause we clone the changes before we change them right? so there is no need to bring them back to the prev state

**Review by edenhassid — COMMENTED** [review]
looks good! added some comments

File: packages/pagerduty-adapter/src/adapter_creator.ts (hunk 1 above)
Comment (Uri-Bar) [reply packages/pagerduty-adapter/src/adapter_creator.ts:74]: talkef f2f

File: packages/pagerduty-adapter/src/filters/users.ts (hunk 5 above)
Comment (Uri-Bar) [reply packages/pagerduty-adapter/src/filters/users.ts:115]: 1. talked f2f
2. fixed

File: packages/pagerduty-adapter/src/filters/users.ts (hunk 6 above)
Comment (Uri-Bar) [reply packages/pagerduty-adapter/src/filters/users.ts:155]: I think its just one liner, if it ok I will leave it her.

File: packages/pagerduty-adapter/src/filters/users.ts (hunk 7 above)
Comment (Uri-Bar) [reply packages/pagerduty-adapter/src/filters/users.ts:158]: removing the on deploy broke the flow, so its stays

**Review by edenhassid — APPROVED** [review]
great!

**Pull request merged**